### PR TITLE
Rename `Osin` CRD to `AuthenticationOperatorConfig`

### PR DESCRIPTION
**Pull request by stlaz**
This renames the `Osin` CRD to `AuthenticationOperatorConfig` and moves it to the `authentication.operator.openshift.io` API group. It also tries to make it cluster-based instead of namespaced.

I'm pretty sure there are pieces I forgot about + I did not test it yet so marking as WIP.

### CODE REVIEW

File: manifests/0000_09_cluster-osin-operator_02_crd.yaml [hunk 1]
```diff
@@ -1,13 +1,13 @@
 apiVersion: apiextensions.k8s.io/v1beta1
 kind: CustomResourceDefinition
 metadata:
-  name: osins.osin.openshift.io
+  name: authenticationoperatorconfig.authentication.operator.openshift.io
```
Comment (enj) [inline manifests/0000_09_cluster-osin-operator_02_crd.yaml]: this must be <plural>.<group>, you are missing an `s`

File: manifests/0000_09_cluster-osin-operator_02_crd.yaml [hunk 2]
```diff
@@ -1,13 +1,13 @@
...
-  group: osin.openshift.io
+  group: authentication.operator.openshift.io
   names:
-    kind: Osin
-    listKind: OsinList
-    plural: osins
-    singular: osin
-  scope: Namespaced
+    kind: AuthenticationOperatorConfig
+    listKind: AuthenticationOperatorConfigList
```
Comment (enj) [inline manifests/0000_09_cluster-osin-operator_02_crd.yaml]: omit

File: manifests/0000_09_cluster-osin-operator_02_crd.yaml [hunk 3]
```diff
@@ -1,13 +1,13 @@
...
+    plural: authenticationoperatorconfigs
+    singular: authenticationoperatorconfig
+  scope: Cluster
```
Comment (enj) [inline manifests/0000_09_cluster-osin-operator_02_crd.yaml]: move this to the top, and in general, make this look exactly like 

```yaml
apiVersion: apiextensions.k8s.io/v1beta1
kind: CustomResourceDefinition
metadata:
  name: kubeapiserveroperatorconfigs.kubeapiserver.operator.openshift.io
spec:
  scope: Cluster
  group: kubeapiserver.operator.openshift.io
  version: v1alpha1
  names:
    kind: KubeAPIServerOperatorConfig
    plural: kubeapiserveroperatorconfigs
    singular: kubeapiserveroperatorconfig
    categories:
    - coreoperators
  subresources:
    status: {}
```

i.e. same categories, need status, etc

File: pkg/apis/authentication/v1alpha1/types.go [hunk 4]
```diff
@@ -9,27 +9,27 @@ import (
 // +genclient
```
Comment (enj) [inline pkg/apis/authentication/v1alpha1/types.go]: You need some magic to tell it the scope of the resource.

```
// +genclient
// +genclient:nonNamespaced
// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
```

File: pkg/operator/operator.go [hunk 5]
```diff
@@ -24,28 +24,28 @@ const (
...
 
-type osinOperator struct {
+type authOperator struct {
 	configMap                   coreclientv1.ConfigMapsGetter
 	kubeAPIServerOperatorConfig dynamic.ResourceInterface
 }
 
-func NewOsinOperator(cmi v1.ConfigMapInformer, cm coreclientv1.ConfigMapsGetter, operatorConfigInformer controller.InformerGetter, operatorConfig dynamic.ResourceInterface) operator.Runner {
-	c := &osinOperator{
+func NewAuthenticationOperator(cmi v1.ConfigMapInformer, cm coreclientv1.ConfigMapsGetter, operatorConfigInformer controller.InformerGetter, operatorConfig dynamic.ResourceInterface) operator.Runner {
```
Comment (enj) [inline pkg/operator/operator.go]: Do not touch this file.  It works and keeps `kube:admin` functioning.  Please leave it alone.  We will delete it when we are ready.

File: pkg/operator2/starter.go [hunk 6]
```diff
@@ -82,13 +80,13 @@ func RunOperator(ctx *controllercmd.ControllerContext) error {
 
 	v1helpers.EnsureOperatorConfigExists(
 		dynamicClient,
-		[]byte(osinResource),
-		osinv1alpha1.GroupVersion.WithResource("osins"),
+		[]byte(authConfigResource),
+		authv1alpha1.GroupVersion.WithResource("osins"),
```
Comment (enj) [inline pkg/operator2/starter.go]: Fix

File: pkg/operator2/starter.go [hunk 7]
```diff
@@ -82,13 +80,13 @@ func RunOperator(ctx *controllercmd.ControllerContext) error {
...
 	)
 
-	operator := NewOsinOperator(
-		osinInformersNamespaced.Osin().V1alpha1().Osins(),
-		osinClient.OsinV1alpha1(),
+	operator := NewAuthenticationOperator(
+		authConfigInformers.Authentication().V1alpha1().AuthenticationOperatorConfigs(),
+		authConfigClient.AuthenticationV1alpha1(), // TODO: Why does update-codegen.sh still generate Osin stuff here?
```
Comment (enj) [inline pkg/operator2/starter.go]: Where?

File: pkg/operator2/starter.go [hunk 8]
```diff
@@ -14,22 +14,21 @@ import (
...
-apiVersion: osin.openshift.io/v1alpha1
-kind: Osin
+	authConfigResource = `
+apiVersion: authentication.operator.openshift.io/v1alpha1
+kind: AuthenticationOperatorConfig
 metadata:
   name: openshift-osin
-  namespace: openshift-osin
 spec:
   managementState: Managed
```
Comment (enj) [inline pkg/operator2/starter.go]: Change to Paused

File: pkg/operator2/starter.go [hunk 9]
```diff
@@ -14,22 +14,21 @@ import (
...
 	resync = 20 * time.Minute
 
-	osinResource = `
-apiVersion: osin.openshift.io/v1alpha1
-kind: Osin
+	authConfigResource = `
+apiVersion: authentication.operator.openshift.io/v1alpha1
+kind: AuthenticationOperatorConfig
 metadata:
   name: openshift-osin
```
Comment (enj) [inline pkg/operator2/starter.go]: Rename to `cluster` and fix where we `Get` it.

File: pkg/operator2/operator.go [hunk 10]
```diff
@@ -99,18 +99,18 @@ func NewOsinOperator(
 	)
 }
 
-func (c *osinOperator) Key() (metav1.Object, error) {
-	return c.osin.Get(targetName, metav1.GetOptions{})
+func (c *authOperator) Key() (metav1.Object, error) {
+	return c.authConfig.Get(targetName, metav1.GetOptions{})
```
Comment (enj) [inline pkg/operator2/operator.go]: change this to Get `cluster`

File: pkg/operator/starter.go [hunk 11]
```diff
@@ -47,7 +47,7 @@ func RunOperator(ctx *controllercmd.ControllerContext) error {
 	kubeAPIServerOperatorConfig := dynamicClient.Resource(kubeAPIServerOperatorConfigGVR)
 	kubeAPIServerOperatorConfigInformer := dynamicInformer(kubeAPIServerOperatorConfig, targtKubeAPIServerOperatorConfig)
 
-	operator := NewOsinOperator(
+	operator := NewAuthenticationOperator(
```
Comment (enj) [inline pkg/operator/starter.go]: undo

**Comment by enj** [comment]
When you have everything sorted in this PR, add a commit to revert f4155479d98fecad8ad7fcf5da3a02d983d1a9e7

File: pkg/operator2/starter.go (hunk 7 above)
Comment (stlaz) [reply pkg/operator2/starter.go]: Ah, I managed to resolve that but left the note there, my bad.

File: pkg/apis/authentication/v1alpha1/types.go (hunk 4 above)
Comment (stlaz) [reply pkg/apis/authentication/v1alpha1/types.go]: Thanks, I was wondering something would probably have to be changed here in order to get this non-namespaced as I went through the generated code, but I have literally no knowledge about these directives.

**Comment by stlaz** [comment]
Thank you for the review!

All should be fixed. I _think_ there was only that one `Get()` where we were using the name of the operator config resource that's now cluster-wide.

**Comment by stlaz** [comment]
... added a newline at the end of the yaml file

File: pkg/operator2/operator.go [hunk 12]
```diff
@@ -84,8 +86,8 @@ func NewOsinOperator(
 	osinNameFilter := operator.FilterByNames(targetName)
 	configNameFilter := operator.FilterByNames(configName)
 
-	return operator.New("OsinOperator2", c,
-		operator.WithInformer(osinInformer, osinNameFilter),
+	return operator.New("AuthenticationOperator2", c,
+		operator.WithInformer(authConfigInformer, osinNameFilter),
```
Comment (enj) [inline pkg/operator2/operator.go]: Name filter is wrong for this one now.

File: pkg/operator2/starter.go [hunk 13]
```diff
@@ -66,9 +65,8 @@ func RunOperator(ctx *controllercmd.ControllerContext) error {
 		informers.WithTweakListOptions(singleNameListOptions(targetName)),
 	)
 
-	osinInformersNamespaced := osininformer.NewSharedInformerFactoryWithOptions(osinClient, resync,
-		osininformer.WithNamespace(targetName),
-		osininformer.WithTweakListOptions(singleNameListOptions(targetName)),
+	authConfigInformers := authopinformer.NewSharedInformerFactoryWithOptions(authConfigClient, resync,
+		authopinformer.WithTweakListOptions(singleNameListOptions(targetName)),
```
Comment (enj) [inline pkg/operator2/starter.go]: Wrong target.

**Comment by stlaz** [comment]
Fixed latest comments. Should have gotten this right the first time, sorry,

**Pull request merged**